File: frontend/src/types/inventory22.ts
/** Domain types for inventory / InventoryRecord22 */
export type InventoryRecord22Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface InventoryRecord22 {
  id: string;
  externalCode: string;
  displayName: string;
  status: InventoryRecord22Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface InventoryRecord22CreateRequest {
  externalCode: string;
  displayName: string;
  status: InventoryRecord22Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface InventoryRecord22ListResponse {
  items: InventoryRecord22[];
  total?: number;
}

export interface InventoryRecord22StatsResponse {
  activeCount: number;
}

export function isInventoryRecord22Active(rec: InventoryRecord22): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatInventoryRecord22Label(rec: InventoryRecord22): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareInventoryRecord22ByPriority(a: InventoryRecord22, b: InventoryRecord22): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/inventory_22.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// InventoryRecord22Service encapsulates business rules for InventoryRecord22.
type InventoryRecord22Service struct {
	repo *repository.InventoryRecord22Repository
}

// NewInventoryRecord22Service wires the service.
func NewInventoryRecord22Service(repo *repository.InventoryRecord22Repository) *InventoryRecord22Service {
	return &InventoryRecord22Service{repo: repo}
}

// Create validates and persists a new InventoryRecord22.
func (s *InventoryRecord22Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.InventoryRecord22, error) {
	now := time.Now().UTC()
	rec := &models.InventoryRecord22{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a InventoryRecord22 by id.
func (s *InventoryRecord22Service) Get(ctx context.Context, id string) (*models.InventoryRecord22, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped InventoryRecord22 pages.
func (s *InventoryRecord22Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.InventoryRecord22, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing InventoryRecord22.
func (s *InventoryRecord22Service) Update(ctx context.Context, actor string, rec *models.InventoryRecord22) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a InventoryRecord22.
func (s *InventoryRecord22Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds InventoryRecord22 by external code prefix.
func (s *InventoryRecord22Service) Search(ctx context.Context, facilityID, prefix string) ([]models.InventoryRecord22, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *InventoryRecord22Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
